src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,466 @@
+            timestamp.Value.GetTimeBucketKey(),
+            _ => new ConcurrentDictionary<string, Metric>()
+        );
+
+        timeBucket.AddOrUpdate(
+            GetMetricBucketKey(type, key, unit.Value, tags),
+            addValuesFactory,
+            (_, metric) =>
+            {
+                // This prevents multiple threads from trying to mutate the metric at the same time. The only other
+                // operations performed against metrics are adding one to the bucket (guaranteed to be atomic due to
+                // the use of a ConcurrentDictionary for the timeBucket) and removing buckets entirely.
+                //
+                // With a very small flushShift (e.g. 0.0) it might be possible for a metric to be emitted to a bucket
+                // that was removed after a flush, in which case that metric.Add(value) would never make it to Sentry.
+                // We've never seen this happen in unit testing (where we always set the flushShift to 0.0) so this
+                // remains only a theoretical possibility of data loss (not confirmed). If this becomes a real problem
+                // and we need to guarantee delivery of every metric.Add, we'll need to build a more complex mechanism
+                // to coordinate flushing with emission.
+                lock(metric)
+                {
+                    metric.Add(value);
+                }
+                return metric;
+            });
+
+        if (_options.ExperimentalMetrics is { EnableCodeLocations: true })
+        {
+            RecordCodeLocation(type, key, unit.Value, stackLevel + 1, timestamp.Value);
+        }
+    }
+
+    private readonly SemaphoreSlim _codeLocationLock = new(1,1);
+
+    internal void RecordCodeLocation(
+        MetricType type,
+        string key,
+        MeasurementUnit unit,
+        int stackLevel,
+        DateTime timestamp
+    )
+    {
+        var startOfDay = timestamp.GetDayBucketKey();
+        var metaKey = new MetricResourceIdentifier(type, key, unit);
+
+        var seenToday = _seenLocations.GetOrAdd(startOfDay,_ => []);
+        if (seenToday.Contains(metaKey))
+        {
+            return;
+        }
+        _codeLocationLock.Wait();
+        try
+        {
+            // Group metadata by day to make flushing more efficient.
+            seenToday.Add(metaKey);
+            if (GetCodeLocation(stackLevel + 1) is not { } location)
+            {
+                return;
+            }
+
+            if (!_pendingLocations.ContainsKey(startOfDay))
+            {
+                _pendingLocations[startOfDay] = new Dictionary<MetricResourceIdentifier, SentryStackFrame>();
+            }
+            _pendingLocations[startOfDay][metaKey] = location;
+        }
+        finally
+        {
+            _codeLocationLock.Release();
+        }
+    }
+
+    internal SentryStackFrame? GetCodeLocation(int stackLevel)
+    {
+        var stackTrace = new StackTrace(true);
```

Comment:
this is a rather expensive API so i wonder if we can leverage https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callermembernameattribute?view=net-8.0 instead

Reply:
I don't see an easy way to do so. There's no way to supply a stackLevel to the `CallerMemberName`, that I can see.